### ISSUE TITLE
MC-147729: Can't find any space for item in the inventory

### ISSUE THREAD

**Issue by mibby** [PaperMC/Paper#1766]
Paper dev 488 (Spigot 1.13.2)

For the last two months, I've been trying to narrow down and deal with a very dangerous and ugly server crash caused by ```Can't find any space for item in the inventory```.

I'm not 100% certain where it comes from but it's part of one of the following;
* A bad chest chunk conversion.
* A bad 1.12 schematic pasting a build with chests next to each other in a 1.13 environment.
* Broken virtual container logic no longer able to understand inventory slots in 1.13.

This error never occurred in 1.12.2, only after I updated to 1.13+. When it occurs, it spams the message repeatedly non-stop thousands of times a second until the server lags to death or eventually crashes. The log file manages to generate ~500MB a minute.

```
[09:12:40] [Server thread/ERROR]: Can't find any space for item in the inventory
[09:12:40] [Server thread/ERROR]: Can't find any space for item in the inventory
[09:12:40] [Server thread/ERROR]: Can't find any space for item in the inventory
[09:12:40] [Server thread/ERROR]: Can't find any space for item in the inventory
[09:12:40] [Server thread/ERROR]: Can't find any space for item in the inventory
[09:12:40] [Server thread/ERROR]: Can't find any space for item in the inventory
[09:12:40] [Server thread/ERROR]: Can't find any space for item in the inventory
[09:12:40] [Server thread/ERROR]: Can't find any space for item in the inventory
[09:12:40] [Server thread/ERROR]: Can't find any space for item in the inventory
[09:12:40] [Server thread/ERROR]: Can't find any space for item in the inventory
[09:12:40] [Server thread/ERROR]: Can't find any space for item in the inventory
[09:12:40] [Server thread/ERROR]: Can't find any space for item in the inventory
[09:12:40] [Server thread/ERROR]: Can't find any space for item in the inventory
[09:12:40] [Server thread/ERROR]: Can't find any space for item in the inventory
[09:12:40] [Server thread/ERROR]: Can't find any space for item in the inventory
[09:12:40] [Server thread/ERROR]: Can't find any space for item in the inventory
[09:12:40] [Server thread/ERROR]: Can't find any space for item in the inventory
[09:12:40] [Server thread/ERROR]: Can't find any space for item in the inventory
```

There is no indication what it comes from before the message but I've had the most reports of it happening when pasting in schematics with a chest in the build.

Would it be possible to protect against crashes / lag / error spam when there is not any space for an item in the inventory? Such as trying to find space and if it can't, deny / delete the item to prevent destruction?

**Comment by Black-Hole** [comment]
That message is logged if a player is using a auto recipe and the inventory can't store the resulting items. If you can prove any correlation to a crash, please provide some logs.

Anyway the message could be changed to include the player using the auto recipe.

**Comment by mibby** [comment]
Unfortunately there is no crash log. The main thread just eventually freezes if it doesn't outright break due to an OOME or no disk space from the error spamming.

If it could include the player using the auto recipe and perhaps _what_ recipe is being used, that would certainly be helpful to diagnosing the true cause of the error.

**Comment by zachbr** [comment]
At this point I think we just need to get additional information into this warning and then we can proceed from there. That seems to be the conclusion from earlier here as well.

**Comment by Strahilchu** [comment]
I just had this error spam my console and came here to see if anyone knew how to prevent it

**Comment by aikar** [comment]
forgot to link it in commit, fixed in latest 1.15.2 (388+) and 1.16.3 (224+)
